### PR TITLE
Deploy OVN emulation with docker and ansible

### PR DESCRIPTION
**Pull request by huikang**
Signed-off-by: Hui Kang kangh@us.ibm.com

### CODE REVIEW

File: ansible/README.rst [hunk 1]
```diff
@@ -0,0 +1,30 @@
+Ansible and Docker OVN Emulation Guide
```
Comment (l8huang) [inline ansible/README.rst:1]: Give a name to your tool, create a folder  move 'ansible' and 'docker' folder into it. Keeping them together can avoid confuse others.

File: ansible/README.rst [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+======================================
+
+Overview
+--------
+
+Deploy an OVN emulation environment using Docker container and ansible
```
Comment (l8huang) [inline ansible/README.rst:7]: This tool can deploy an environment using Docker and ansible, how to run test cases on this environment, eg. create ports, please give a description about it.

File: ansible/roles/ovn/defaults/main.yml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+---
+project_name: "ovn"
+
+#######################
+# Docker
+#######################
+
+ovn_db_image_full: "huikang/ovn-scale-test"
+ovn_chassis_image_full: "huikang/ovn-scale-test"
+
```
Comment (l8huang) [inline ansible/roles/ovn/defaults/main.yml:10]: Avoid specific username in configure file, replace with a common placeholder.

File: ansible/roles/ovn/tasks/config.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
+---
```
Comment (l8huang) [inline ansible/roles/ovn/tasks/config.yml]: I didn't use ansible before, don't know how it works:)

File: ansible/roles/ovn/tasks/start.yml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+- name: start OVN chassis
+  docker:
+    name: "ovn-chassis-{{ item }}"
+    image: "{{ ovn_db_image_full }}"
+    pull: always
+    privileged: yes
+    net: host
+    detach: True
+    state: started
+    command: "ovn_set_chassis --ovn --controller-ip {{ ovn_database_alias_ip}} --host-ip {{ ovn_chassis_start_cidr_ip.split('.')[0] }}.{{ ovn_chassis_start_cidr_ip.split('.')[1] }}.{{ ovn_chassis_start_cidr_ip.split('.')[2] | int + groups['emulation-hosts'].index(inventory_hostname) | int }}.{{ ovn_chassis_start_cidr_ip.split('.')[3] | int + item | int }}/{{ ovn_chassis_cidr_mask }} --device {{ ovn_database_device }}"
```
Comment (l8huang) [inline ansible/roles/ovn/tasks/start.yml]: Seems like a IP generator function is needed here, current way is not safe.

File: docker/Makefile [hunk 6]
```diff
@@ -0,0 +1,4 @@
+image_name=huikang/ovn-scale-test
```
Comment (l8huang) [inline docker/Makefile]: Replace your name with a common path name ;)

File: ansible/README.rst (hunk 1 above)
Comment (huikang) [inline ansible/README.rst:1]: What do you think a folder named "ansible-docker"? I will move all changes under this new folder.

File: ansible/README.rst (hunk 2 above)
Comment (huikang) [inline ansible/README.rst:7]: Once deployed with ansible, it is supposed to setup rally environment like

```
rally deployment create --file=ovn-emulation.json --name=ovn-emulation
```

So once this PR got cleaned up, I will send another PR that extends rally to understand the OVN emulation environment. Of course, a large portion of this has done by the rally modified by you.

File: ansible/roles/ovn/defaults/main.yml (hunk 3 above)
Comment (huikang) [inline ansible/roles/ovn/defaults/main.yml:10]: Would you mind creating an account in docker hub, like "ovn-scale-test"? Then I can upload the image to that repository and avoid using this specific name.

I will update the README about how to use customized image.

File: ansible/roles/ovn/tasks/start.yml (hunk 5 above)
Comment (huikang) [inline ansible/roles/ovn/tasks/start.yml]: Agree. I will write an ansible python module that can take input of the start_cird and generate a list of cidr for the fake chassis. - Hui

File: docker/Makefile (hunk 6 above)
Comment (huikang) [inline docker/Makefile]: see my reply above. Thanks. - Hui

File: ansible/roles/ovn/tasks/start.yml (hunk 5 above)
Comment (huikang) [inline ansible/roles/ovn/tasks/start.yml]: The IP generate ansible module is added in the updated patch.

File: ansible/README.rst (hunk 1 above)
Comment (l8huang) [inline ansible/README.rst:1]: The name is ok.

**Comment by l8huang** [comment]
Hi Hui, when will you send a updated PR, I have no more comments for the current PR. Appreciate your works :)

**Comment by huikang** [comment]
Lei, let me add more details in the README file. I will send the update by today. After getting merged, I will work on the Rally part. Thanks. - Hui

**Comment by huikang** [comment]
Hi, Lei, could you merge this updated PR? Please let me know if you have any question. Thanks. - Hui

**Pull request merged**